feat(faucet): new faucet

jlehtimaki opened the pull request:
This version of the Faucet does not have any limitations per user, only limitations are daily token supply.
Faucet will have the following features:
- New UI for sending tokens
- Spinner to wait for TX to go through
- Display TX when TX is confirmed
- Sending tokens by batch, does not even try to send them immediately. The send time is configurable.
- Setting up configuration by env or file

It uses HTMX + JS.

You can run it with simply:
```
go run .
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
  - Introduced a new web-based Faucet service for token distribution.

- **Improvements**
  - Enhanced naming conventions for better clarity with the introduction of `FulfilKeyRequest` and `SignRequest`.

- **Documentation**
  - Updated `CHANGELOG.md` with new version details and module changes.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jlehtimaki committed:
feat(faucet): new faucet

This version of the Faucet does not have any limitations per user, only
limitations are daily token supply

jjheywood committed:
fix: minor css changes for mobile

jjheywood reviewed (commented):
Made some minor css updates for mobile but apart from that its all good from my side

pitasi commented on cmd/faucet/assets/background.jpg:
can we compress this image? I think it should be resized to a reasonable size (since it's being rendered at a small size anyway, and potentially either converted to webp or to a lower jpeg quality (usually 60-65 is sufficient)

jlehtimaki replied:
Compressed

jlehtimaki committed:
fix(faucet): add daily refresh and token refresh job

jlehtimaki committed:
fix(faucet): fix bugs and ci

jlehtimaki committed:
fix(faucet): remove unnessary go.mod files, replace dockerfile faucet-v2

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

pitasi commented on cmd/faucet/faucet.go:
```diff
@@ -178,288 +134,164 @@ func validAddress(addr string) error {
+	var err error
+	cfg, err := config.LoadConfig()
+	if err != nil {
+		panic(err)
```
i think the bot was right though, any reason for panicking here?

jlehtimaki replied:
The idea was that if the configuration is incorrect, there's no reason for the process to start.
I'll change it to Fatal logging.

pitasi commented on cmd/faucet/pkg/config/config.go:
```diff
@@ -0,0 +1,85 @@
+	CoolDown       string `env:"COOLDOWN" envDefault:"10s" mapstructure:"COOLDOWN"`
+	KeyringBackend string `env:"KEYRING" envDefault:"test" mapstructure:"KEYRING"`
+	BatchInterval  string `env:"BATCH_INTERVAL" envDefault:"10s" mapstructure:"BATCH_INTERVAL"`
+	DailyLimit     string `env:"DAILY_LIMIT" envDefault:"100000000" mapstructure:"DAILY_LIMIT"`
```
can this be an int?

pitasi commented on cmd/faucet/pkg/config/config.go:
```diff
@@ -0,0 +1,85 @@
+	Fees           string `env:"FEES" envDefault:"25uward" mapstructure:"FEES"`
+	CoolDown       string `env:"COOLDOWN" envDefault:"10s" mapstructure:"COOLDOWN"`
+	KeyringBackend string `env:"KEYRING" envDefault:"test" mapstructure:"KEYRING"`
+	BatchInterval  string `env:"BATCH_INTERVAL" envDefault:"10s" mapstructure:"BATCH_INTERVAL"`
```
can this be a time.Duration?

pitasi commented on cmd/faucet/pkg/config/config.go:
```diff
@@ -0,0 +1,85 @@
+	Denom          string `env:"DENOM" envDefault:"uward" mapstructure:"DENOM"`
+	Amount         string `env:"AMOUNT" envDefault:"1000000" mapstructure:"AMOUNT"`
+	Fees           string `env:"FEES" envDefault:"25uward" mapstructure:"FEES"`
+	CoolDown       string `env:"COOLDOWN" envDefault:"10s" mapstructure:"COOLDOWN"`
```
can this be a time.Duration?

jlehtimaki replied:
We don't actually need the cooldown for this

pitasi commented on cmd/faucet/faucet.go:
```diff
@@ -178,288 +134,164 @@ func validAddress(addr string) error {
-	}, " ")
+func MutexLocked(m *sync.Mutex) bool {
+	state := reflect.ValueOf(m).Elem().FieldByName("state")
+	return state.Int()&mutexLocked == mutexLocked
```
i think that generally speaking this is a very bad practice, as this method itself is not thread safe 😁 

MutexLocked could return `false`. But immediately after the mutex might get locked.

If you accept this, it's okay. My comment is to acknowledge that this isn't really a protection and two f.Send() might still occur simultaneously.

jlehtimaki replied:
Got it, removed the lock check

pitasi commented on cmd/faucet/faucet.go:
```diff
@@ -178,288 +134,164 @@ func validAddress(addr string) error {
-	defer l.mu.Unlock()
-	for _, k := range key {
-		l.last[k] = time.Time{}
+		f.TokensAvailable = f.DailySupply
```
you should acquire the lock otherwise there is a race condition between this and f.Send() trying to read/write this var

jlehtimaki replied:
Added lock

jlehtimaki committed:
chore(faucet): remove unnessary variables and modify types

this will modify the types of DailyLimit and Batchinterval to remove
unnessary conversions and remove Cooldown and BatchInterval

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

jlehtimaki committed:
fix(faucet): acquire a lock when setting the new supply

jlehtimaki committed:
fix(faucet): removed locking where it's not needed

other enhancements as well, the loading and setting context to correct
place

jlehtimaki committed:
fix(faucet): converted background to much smaller image

jlehtimaki committed:
feat(faucet): make the run out of daily nicer

pitasi reviewed (dismissed):
the code lgtm but beware that the same address can be added multiple times to a batch, so i can do a "while true; curl" to obtain all the daily funds pretty easily 😁 

```
2024-07-26T11:43:15+02:00 INF sending 1000000uward to [warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv warden1ucgwsdqsdva08095erm7a02lj6pzwx0m03w2kv]
```

jlehtimaki committed:
Merge branch 'main' of github.com:warden-protocol/wardenprotocol into new-new-faucet

jlehtimaki committed:
feat(faucet): add new style

jlehtimaki committed:
feat(faucet): add check that same address cannot be multiple time in one batch

jlehtimaki committed:
feat(faucet): add metrics

jlehtimaki committed:
fix(faucet): Dockerfile libwasmvm was missing access rights

jlehtimaki committed:
feat(faucet): add pre input option to address form

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

jlehtimaki committed:
Merge branch 'main' into new-new-faucet

jlehtimaki committed:
Merge branch 'main' into new-new-faucet
